File: app/src/main/java/com/example/trigger/https/HttpsTools.java
package com.example.trigger.https;

import android.util.Base64;

import java.io.ByteArrayInputStream;
import java.security.KeyManagementException;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.security.cert.Certificate;
import java.security.cert.CertificateException;
import java.security.cert.CertificateFactory;
import java.security.cert.X509Certificate;

import javax.net.ssl.HostnameVerifier;
import javax.net.ssl.HttpsURLConnection;
import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLSession;
import javax.net.ssl.TrustManager;
import javax.net.ssl.X509TrustManager;

import com.example.trigger.Log;


public class HttpsTools {

    public static boolean isValid(X509Certificate cert) {
        try {
            cert.checkValidity();
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    public static boolean isSelfSigned(X509Certificate cert) {
        return cert.getIssuerX500Principal().getName().equals(
            cert.getSubjectX500Principal().getName()
        );
    }

    public static String serializeCertificate(Certificate cert) {
        if (cert == null) {
            return "";
        }

        try {
            return "-----BEGIN CERTIFICATE-----\n"
                + Base64.encodeToString(cert.getEncoded(), Base64.DEFAULT)
                + "\n-----END CERTIFICATE-----\n";
        } catch (Exception e) {
            Log.e("HttpsTools", e.toString());
        }

        return "";
    }

    public static Certificate deserializeCertificate(String cert) {
        if (cert == null || cert.length() == 0) {
            return null;
        }

        try {
            ByteArrayInputStream derInputStream = new ByteArrayInputStream(cert.getBytes());
            CertificateFactory certificateFactory = CertificateFactory.getInstance("X.509"); //KeyStore.getDefaultType() => "BKS"
            return certificateFactory.generateCertificate(derInputStream);
        } catch (Exception e) {
            Log.e("HttpsTools", e.toString());
        }

        return null;
    }

    // disable any certificate validation
    public static void disableDefaultHostnameVerifier() {
        HttpsURLConnection.setDefaultHostnameVerifier(new HostnameVerifier() {
            @Override
            public boolean verify(String arg0, SSLSession arg1) {
                return true;
            }
        });
    }

    // disable any certificate validation
    public static void disableDefaultCertificateValidation()
            throws NoSuchAlgorithmException, KeyManagementException {
        TrustManager trustManager = new X509TrustManager() {
            @Override
            public void checkClientTrusted(X509Certificate[] cert, String authType)
                    throws CertificateException {
            }

            @Override
            public void checkServerTrusted(X509Certificate[] cert, String authType)
                    throws CertificateException {
            }

            @Override
            public X509Certificate[] getAcceptedIssuers() {
                return new X509Certificate[]{};
            }
        };

        TrustManager[] trustManagers = new TrustManager[]{ trustManager };
        SSLContext context = SSLContext.getInstance("TLS");
        context.init(null, trustManagers, new SecureRandom());
        HttpsURLConnection.setDefaultSSLSocketFactory(context.getSocketFactory());
    }
}

File: app/src/main/kotlin/app/trigger/https/HttpsTools.kt
/*
* Copyright (C) 2025 The Trigger Contributors
* SPDX-License-Identifier: GPL-3.0-or-later
*/

package app.trigger.https

import kotlin.Throws
import android.util.Base64
import app.trigger.Log
import java.io.ByteArrayInputStream
import java.lang.Exception
import java.security.*
import java.security.cert.Certificate
import java.security.cert.CertificateException
import java.security.cert.CertificateFactory
import java.security.cert.X509Certificate
import javax.net.ssl.*

object HttpsTools {
    private const val TAG = "HttpsTools"
    fun isValid(cert: X509Certificate): Boolean {
        return try {
            cert.checkValidity()
            true
        } catch (e: Exception) {
            false
        }
    }

    fun isSelfSigned(cert: X509Certificate): Boolean {
        return cert.issuerX500Principal.name ==
                cert.subjectX500Principal.name
    }

    fun serializeCertificate(certificate: Certificate?): String {
        if (certificate == null) {
            return ""
        }
        try {
            val prefix = "-----BEGIN CERTIFICATE-----"
            val mid = Base64.encodeToString(certificate.encoded, Base64.DEFAULT)
            val suffix = "-----END CERTIFICATE-----"
            return "${prefix}\n${mid}\n${suffix}"
        } catch (e: Exception) {
            Log.e(TAG, e.toString())
        }
        return ""
    }

    fun deserializeCertificate(certificate: String?): Certificate? {
        if (certificate.isNullOrEmpty()) {
            return null
        }
        try {
            val derInputStream = ByteArrayInputStream(certificate.toByteArray())
            val certificateFactory = CertificateFactory.getInstance("X.509") //KeyStore.getDefaultType() => "BKS"
            return certificateFactory.generateCertificate(derInputStream)
        } catch (e: Exception) {
            Log.e(this, e.toString())
        }
        return null
    }

    // disable any certificate validation
    fun disableDefaultHostnameVerifier() {
        HttpsURLConnection.setDefaultHostnameVerifier { arg0, arg1 -> true }
    }

    // disable any certificate validation
    @Throws(NoSuchAlgorithmException::class, KeyManagementException::class)
    fun disableDefaultCertificateValidation() {
        val trustManager: TrustManager = object : X509TrustManager {
            @Throws(CertificateException::class)
            override fun checkClientTrusted(cert: Array<X509Certificate>, authType: String) {
            }

            @Throws(CertificateException::class)
            override fun checkServerTrusted(cert: Array<X509Certificate>, authType: String) {
            }

            override fun getAcceptedIssuers(): Array<X509Certificate> {
                return arrayOf()
            }
        }
        val trustManagers = arrayOf(trustManager)
        val context = SSLContext.getInstance("TLS")
        context.init(null, trustManagers, SecureRandom())
        HttpsURLConnection.setDefaultSSLSocketFactory(context.socketFactory)
    }

    @Throws(NoSuchAlgorithmException::class, KeyStoreException::class, KeyManagementException::class)
    fun getSocketFactoryIgnoreCertificateExpiredException(): SSLSocketFactory {
        val factory = TrustManagerFactory.getInstance("X509")
        factory.init(null as KeyStore?)
        val trustManagers = factory.trustManagers
        for (i in trustManagers.indices) {
            if (trustManagers[i] is X509TrustManager) {
                trustManagers[i] = IgnoreExpirationTrustManager(trustManagers[i] as X509TrustManager)
            }
        }
        val sslContext = SSLContext.getInstance("TLS")
        sslContext.init(null, trustManagers, null)
        return sslContext.socketFactory
    }
}
